sno/fast_import.py

```diff
@@ -231,23 +290,34 @@ def fast_import_tables(
                     replacing_dataset = None
 
                 if replacing_dataset is not None:
-                    if replace_ids is None:
-                        # Delete the existing dataset, before we re-import it.
-                        p.stdin.write(f"D {source.dest_path}\n".encode("utf8"))
-                    else:
-                        # delete and reimport meta/
-                        # we also delete the specified features, but we do it further down
-                        # so that we don't have to iterate the IDs more than once.
-                        p.stdin.write(
-                            f"D {source.dest_path}/.sno-dataset/meta\n".encode("utf8")
-                        )
-
-                    # We just deleted the legends, but we still need them to reimport
-                    # data efficiently. Copy them from the original dataset.
-                    for x in write_blobs_to_stream(
-                        p.stdin, replacing_dataset.iter_legend_blob_data()
-                    ):
-                        pass
+                    for i, proc in enumerate(procs):
+                        if replace_ids is None:
+                            # Delete the existing dataset, before we re-import it.
+                            proc.stdin.write(f"D {source.dest_path}\n".encode("utf8"))
+                        else:
+                            # delete and reimport meta/
+                            proc.stdin.write(
+                                f"D {source.dest_path}/.sno-dataset/meta\n".encode(
+                                    "utf8"
+                                )
+                            )
+                            # delete all features not pertaining to this process.
+                            # we also delete the features that *do*, but we do it further down
+                            # so that we don't have to iterate the IDs more than once.
```

Comment:
Oh yeah this is probably something to do with enum ReplaceExisting:
1. DONT_REPLACE importing a brand new dataset you don't currently have (all existing data is retained)
2. GIVEN importing a dataset over the top of a single existing dataset (retaining all other datasets)
3. ALL importing a dataset over the top of the entire repo (no data is retained)

So another way to look at it, if your repo has N datasets, DONT_REPLACE starts the import from a base of N, GIVEN starts from a base of  N-1, and ALL starts from a base of 0.

Which type of ReplaceExisting do you use for mirroring? If it's ALL, then this delete code won't be run anyway for mirroring, so it doesn't make much difference to us... you'd just be cleaning this up for a (hypothetical) user who is using mode GIVEN, which is the only mode that needs to start from somewhere fancy where things need deleting to get to N-1 - starting from N (where we are now) or from 0 (the empty tree) is trivial. Consider switching to ALL if you are not, since it should be slightly cheaper...
 
But Rob is right - perhaps with your new found tree merging skills, we could do some sanity checks, and then we would always do a ReplaceExisting.ALL to import this dataset, merge all its trees into a single tree... and then, merge *that* tree in with the rest of the repo for a ReplaceExisting.GIVEN or DONT_REPLACE. It might even make the code a bit simpler

Reply:
But `--replace-ids` is a list of IDs to reimport. It doesn't cover the whole dataset. The rest of the dataset should be left untouched.